File: Python3/1 - 200/48 Rotate Image.py
# 48ms, 60.17%
class Solution:
    def rotate(self, matrix):
        n = len(matrix)
        r = n // 2
        for i in range(r):
            for j in range(n - 1 - 2 * i):
                tmp = matrix[i][i + j]
                matrix[i][i + j] = matrix[n - i - j - 1][i]
                matrix[n - i - j - 1][i] = matrix[n - i - 1][n - i - j - 1]
                matrix[n - i - 1][n - i - j - 1] = matrix[i + j][n - i - 1]
                matrix[i + j][n - i - 1] = tmp
        

File: 1 - 200/48 Rotate Image.cc
// 5ms, 37.22%
class Solution
{
public:
    void rotate(vector<vector<int>> &matrix)
    {
        int round = matrix.size() / 2;
        while(--round >= 0)
            circle(matrix, round);
    }
    
private:
    void circle(vector<vector<int>> &c, int k)
    {
        int n = c.size() - k - 1;
        for(int i = 0; i < n - k; ++i)
        {
            int tmp = c[k][k + i];
            c[k][k + i] = c[n - i][k];
            c[n - i][k] = c[n][n - i];
            c[n][n - i] = c[k + i][n];
            c[k + i][n] = tmp;
        }
    }
};
